### PR TITLE
Grid Master Detail

### LINKED ISSUE #5880: Allow master-detail style visualization with igxGrid
## Is your feature request related to a problem? Please describe.
I want to be able to create a master-detail view with igxGrid, where the master is the grid rows that are expandable and expanding them shows custom template / component for their details.

## Describe the solution you'd like
Clear-cut way for defining "details" template for grid rows.

## Describe alternatives you've considered
N/A

## Additional context
N/A

### PR DESCRIPTION
**Pull request by MayaKirova**
Closes #5880  

To do:
- [x] Finalize the spec - https://github.com/IgniteUI/igniteui-angular/wiki/Grid-Master-Detail
- [x] Add test plan
- [x] Automation
- [x] Add sample - https://github.com/IgniteUI/igniteui-angular-samples/pull/1570
- [x] Add documentation. - https://github.com/IgniteUI/igniteui-docfx/pull/1649
- [x] Update the change log.

### Additional information (check all that apply):
 - [ ] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code
 - [ ] This PR includes API docs for newly added methods/properties
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.ts [hunk 1]
```diff
@@ -527,6 +535,215 @@ export class IgxGridComponent extends IgxGridBaseDirective implements GridType,
...
+    private cloneMap(mapIn: Map<any, boolean>): Map<any, boolean> {
+        const mapCloned: Map<any, boolean> = new Map<any, boolean>();
+
+        mapIn.forEach((value: boolean, key: any, mapObj: Map<any, boolean>) => {
+
+            mapCloned.set(key, value);
+        });
+
+        return mapCloned;
+    }
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.ts]: Is this method really needed?

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.ts [hunk 2]
```diff
@@ -527,6 +535,215 @@ export class IgxGridComponent extends IgxGridBaseDirective implements GridType,
...
+     *  <header>{{dataItem.Category?.CategoryName}}</header>
+     * <span>{{dataItem.Category?.Description}}</span>
+     * </div>
+     * </ng-template>
+     * </igx-grid>
+     * ```
+	 * @memberof IgxGridComponent
+     */
+    public set expansionStates(value) {
+        this._expansionStates = this.cloneMap(value);
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.ts]:
```suggestion
        this._expansionStates = new Map<T, U>(value); // or new Map<T, U>(value.entries())
```

File: projects/igniteui-angular/src/lib/grids/cell.component.ts [hunk 3]
```diff
@@ -850,6 +851,26 @@ export class IgxGridCellComponent implements OnInit, OnChanges, OnDestroy {
...
+            this.grid.expansionStates = expandedStates;
+            this.grid.notifyChanges();
+            const isVirtualized = !this.grid.verticalScrollContainer.dc.instance.notVirtual;
+            // persist focused cell
+            const el = this.grid.selectionService.activeElement;
+            if (isVirtualized && el) {
+                const cell = this.grid.gridAPI.get_cell_by_visible_index(el.row, el.column);
+                if (cell) {
+                    cell.nativeElement.focus();
+                }
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/grids/cell.component.ts]: This won't always persist the focus on the right cell element as `notifyChanges` is run on the next detection cycle and the layout of the grid body can shift up/down

Comment (ChronosSF) [reply projects/igniteui-angular/src/lib/grids/cell.component.ts]: Related to #6191

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.ts [hunk 4]
```diff
@@ -527,6 +535,204 @@ export class IgxGridComponent extends IgxGridBaseDirective implements GridType,
...
+            event.stopPropagation();
+            if (shift && target === container) {
+                // shift + tab from details to data row
+                event.preventDefault();
+                const lastColIndex = this.unpinnedColumns[this.unpinnedColumns.length - 1].visibleIndex;
+                this.navigateTo(rowIndex - 1, lastColIndex,
+                    (args) => args.target.nativeElement.focus());
+            } else if (!shift) {
+                const focusable = container
+                .querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])');
```
Comment (ChronosSF) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.ts]: @radomirchev @rkaraivanov 

Alternative to this code we discussed the idea of having a hidden element at the end to take the focus.

However, after discussing this with @MayaKirova , I think a simpler solution with not much worse end-user experience is to focus the outer details container regardless of where the focus is coming from. 

This will look like this:

- Last cell <kbd>Tab</kbd> Details Container  <kbd>Tab</kbd> Details Container contents ....  <kbd>Tab</kbd> First cell
- First cell <kbd>Shift</kbd> + <kbd>Tab</kbd> Details Container <kbd>Shift</kbd> + <kbd>Tab</kbd> Last cell
- First cell <kbd>Shift</kbd> + <kbd>Tab</kbd> Details Container <kbd>Tab</kbd> Details Container contents... 

Basically <kbd>Shift</kbd> + <kbd>Tab</kbd> won't drill down through the details contents, the users will drill down with <kbd>Tab</kbd> in it (or <kbd>Shift</kbd> + <kbd>Tab</kbd> but only while inside)

**Comment by nrobakova** [comment]
Hi,
I tested this PR and found some issue related to it:

- [ ] getSelectedData() returns null values when master detail is in included in the selection

- [ ] Cannot navigate from root summaries to the grid using Shift+Tab keys if the last row is expanded

- [ ] When rowSelection="multiple" and navigate with Tab from master detail. The row selector checkbox is focused, but the first cell in the next row should be focused.

- [ ] When filter and then collapse master detail an error is returned. Steps to reproduce:
     1. Open dev sample gridMasterDetail
     2. Expand first row.
     3. Filter column Contact name with value 'th'
     4. Expand filtered row
     5. Clear filter
     6. Collapse first row
    An error in console is returned

- [ ] When navigate with tab from rowEditng last cell , the master detail is selected instead of the row editing Cancel button
